src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -3015,24 +3042,50 @@ ReceiveResults(WorkerSession *session, bool storeRows)
 	WorkerPool *workerPool = session->workerPool;
 	DistributedExecution *execution = workerPool->distributedExecution;
 	DistributedExecutionStats *executionStats = execution->executionStats;
-	TupleDesc tupleDescriptor = execution->tupleDescriptor;
 	AttInMetadata *attributeInputMetadata = NULL;
 	uint32 expectedColumnCount = 0;
 	char **columnArray = NULL;
 	Tuplestorestate *tupleStore = execution->tupleStore;
+	MemoryContext oldContext = NULL;
+	MemoryContext ioContext = NULL;
 
-	MemoryContext ioContext = AllocSetContextCreate(CurrentMemoryContext,
-													"ReceiveResults",
-													ALLOCSET_DEFAULT_MINSIZE,
-													ALLOCSET_DEFAULT_INITSIZE,
-													ALLOCSET_DEFAULT_MAXSIZE);
-	if (tupleDescriptor != NULL)
-	{
-		attributeInputMetadata = TupleDescGetAttInMetadata(tupleDescriptor);
-		expectedColumnCount = tupleDescriptor->natts;
-		columnArray = (char **) palloc0(expectedColumnCount * sizeof(char *));
-	}
+	MemoryContext receiveResultsContext = AllocSetContextCreate(CurrentMemoryContext,
+																"ReceiveResults",
+																ALLOCSET_DEFAULT_MINSIZE,
+																ALLOCSET_DEFAULT_INITSIZE,
+																ALLOCSET_DEFAULT_MAXSIZE);
+
+	/*
+	 * Allocate all remaining memory within this context. This function may be called many
+	 * times per distributed execution, so we're guarding against any memory leaks with
+	 * using a specific memory context. This context is deleted at the end of the function,
+	 * so be careful to add early returns to the function or relying on anything
+	 * allocated within the function.
+	 *
+	 * Note that tuplestore_puttuple() copies the tuples that are stored in the tuplestore
+	 * in the tuplestore's own memory context. So, we don't need to anything special
+	 * for that.
+	 */
+	oldContext = MemoryContextSwitchTo(receiveResultsContext);
```

Comment:
Is this still used for anything? How much extra overhead does creating a memory context incur?

Reply:
>Is this still used for anything?

No it is not,  see [here](https://github.com/citusdata/citus/pull/3114#discussion_r336956442)

I haven't measured it. But I'm now leaning towards  removing it because this is  a performance critical code-path. Do you also think similar?

Reply:
Yes, let's remove it.